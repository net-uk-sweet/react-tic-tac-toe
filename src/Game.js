import React from 'react';
import Immutable from 'immutable';

import Board from './Board';
import GameInfo from './GameInfo';
import { calculateWinner } from './helpers';
import { defaultState } from './emptyStates';
  
  export default class Game extends React.Component {
  
    constructor(props) {
      super(props);
  
      this.state = this.getInitialState();
    }
  
    getInitialState() {
      return defaultState(this.props.squares, this.props.player);
    }
  
    handleToggleOrderClick() {
      this.setState({
        ascendingOrder: !this.state.ascendingOrder
      })
    }
  
    handleResetClick() {
      this.setState(this.getInitialState());
    }
  
    handleSquareClick(index, row, column) {
      let history = this.state.history.slice(0, this.state.move + 1);
      const current = history.get(history.size - 1);  
      const player = this.state.player ? 0 : 1;

      let squares = current.get('squares');
  
      if (calculateWinner(squares) || squares.getIn([index, 'token'])) {
        return;
      }
  
      squares = squares.set(index, Immutable.fromJS({
        token: this.props.players[this.state.player],
        row,
        column
      }));
      
      history = history.concat(Immutable.fromJS([{ index, squares }]));
  
      this.setState({
        player,
        move: history.size - 1,
        history
      });
    }
  
    handleMoveClick(move) {
      const player = move % 2 ? 1 : 0;
      this.setState({ move, player });
    }
  
    render() {
      const history = this.state.history;
      const current = history.getIn([this.state.move, 'squares']);
      const playerToken = this.props.players[this.state.player];
      const winner = calculateWinner(current);
  
      let status = this.state.move < current.size ? `Next player: ${playerToken}` : 'Draw';
      
      if (winner) {
        status = `Winner: ${winner.get('token')}`;
      }

      return (
        <div className="game">
          <div className="game-board">
            <Board 
              winner={winner && winner.get('sequence')}
              squares={current} 
              columns={this.props.columns}
              onClick={(i, row, column) => { this.handleSquareClick(i, row, column) }} 
            />
          </div>
          <div className="game-panel">
            <GameInfo status={status} 
              history={history} 
              currentMove={this.state.move} 
              ascendingOrder={this.state.ascendingOrder} 
              onMoveClick={(move) => { this.handleMoveClick(move) }}
            />
            <button className="button" onClick={() => { this.handleToggleOrderClick() }}>Toggle order</button>
            <button className="button" onClick={() => { this.handleResetClick() }}>Reset</button>
          </div>
        </div>
      );
    }
  }