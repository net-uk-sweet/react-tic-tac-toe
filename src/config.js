export const GAME = {
    player: 0,
    players: ['x', 'o'],
    squares: 9,
    columns: 3
};