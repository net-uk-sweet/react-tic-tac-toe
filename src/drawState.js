import Immutable from 'immutable';

export const drawState = {
    move: 9,
    ascendingOrder: true,
    history: Immutable.fromJS([
      {
        squares: [
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          }
        ]
      },
      {
        index: 0,
        squares: [
          {
            token: 'x',
            row: 1,
            column: 1
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          }
        ]
      },
      {
        index: 1,
        squares: [
          {
            token: 'x',
            row: 1,
            column: 1
          },
          {
            token: 'o',
            row: 1,
            column: 2
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          }
        ]
      },
      {
        index: 3,
        squares: [
          {
            token: 'x',
            row: 1,
            column: 1
          },
          {
            token: 'o',
            row: 1,
            column: 2
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: 'x',
            row: 2,
            column: 1
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          }
        ]
      },
      {
        index: 4,
        squares: [
          {
            token: 'x',
            row: 1,
            column: 1
          },
          {
            token: 'o',
            row: 1,
            column: 2
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: 'x',
            row: 2,
            column: 1
          },
          {
            token: 'o',
            row: 2,
            column: 2
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          }
        ]
      },
      {
        index: 7,
        squares: [
          {
            token: 'x',
            row: 1,
            column: 1
          },
          {
            token: 'o',
            row: 1,
            column: 2
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: 'x',
            row: 2,
            column: 1
          },
          {
            token: 'o',
            row: 2,
            column: 2
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: 'x',
            row: 3,
            column: 2
          },
          {
            token: null,
            column: null,
            row: null
          }
        ]
      },
      {
        index: 6,
        squares: [
          {
            token: 'x',
            row: 1,
            column: 1
          },
          {
            token: 'o',
            row: 1,
            column: 2
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: 'x',
            row: 2,
            column: 1
          },
          {
            token: 'o',
            row: 2,
            column: 2
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: 'o',
            row: 3,
            column: 1
          },
          {
            token: 'x',
            row: 3,
            column: 2
          },
          {
            token: null,
            column: null,
            row: null
          }
        ]
      },
      {
        index: 5,
        squares: [
          {
            token: 'x',
            row: 1,
            column: 1
          },
          {
            token: 'o',
            row: 1,
            column: 2
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: 'x',
            row: 2,
            column: 1
          },
          {
            token: 'o',
            row: 2,
            column: 2
          },
          {
            token: 'x',
            row: 2,
            column: 3
          },
          {
            token: 'o',
            row: 3,
            column: 1
          },
          {
            token: 'x',
            row: 3,
            column: 2
          },
          {
            token: null,
            column: null,
            row: null
          }
        ]
      },
      {
        index: 8,
        squares: [
          {
            token: 'x',
            row: 1,
            column: 1
          },
          {
            token: 'o',
            row: 1,
            column: 2
          },
          {
            token: null,
            column: null,
            row: null
          },
          {
            token: 'x',
            row: 2,
            column: 1
          },
          {
            token: 'o',
            row: 2,
            column: 2
          },
          {
            token: 'x',
            row: 2,
            column: 3
          },
          {
            token: 'o',
            row: 3,
            column: 1
          },
          {
            token: 'x',
            row: 3,
            column: 2
          },
          {
            token: 'o',
            row: 3,
            column: 3
          }
        ]
      },
      {
        index: 2,
        squares: [
          {
            token: 'x',
            row: 1,
            column: 1
          },
          {
            token: 'o',
            row: 1,
            column: 2
          },
          {
            token: 'x',
            row: 1,
            column: 3
          },
          {
            token: 'x',
            row: 2,
            column: 1
          },
          {
            token: 'o',
            row: 2,
            column: 2
          },
          {
            token: 'x',
            row: 2,
            column: 3
          },
          {
            token: 'o',
            row: 3,
            column: 1
          },
          {
            token: 'x',
            row: 3,
            column: 2
          },
          {
            token: 'o',
            row: 3,
            column: 3
          }
        ]
      }
    ]),
    player: 1
  }