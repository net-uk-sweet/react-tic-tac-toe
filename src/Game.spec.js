import React from 'react';
import { shallow } from 'enzyme';

import { GAME as props } from './config';
import Game from './Game';
import { winState } from './winState';
import { drawState } from './drawState';

let shared = { };

beforeEach(() => {
  shared = {};
  shared.wrapper = shallow(<Game {...props} />);
  shared.initialState = shared.wrapper.state();
});

it('matches snapshot', () => {
  expect(shared.wrapper.getElement()).toMatchSnapshot();
});

it('has the correct default state', () => {
  const squares = shared.initialState.history.getIn([0, 'squares']);
  expect(shared.initialState.history.size).toEqual(1);
  expect(shared.initialState.player).toEqual(0);
  expect(squares.size).toEqual(9);
  expect(squares.get(0).toJS()).toEqual({ token: null, column: null, row: null });
});

it('toggles the move order when the toggle order button is selected', () => {
  shared.wrapper.find('.qa-toggle-button').simulate('click');
  expect(shared.wrapper.state().ascendingOrder).not.toEqual(shared.initialState);
});

it('resets the game when the reset button is selected', () => {
  shared.wrapper.state(winState);
  shared.wrapper.find('.qa-reset-button').simulate('click');
  expect(shared.wrapper.state()).toEqual(shared.initialState);
});

it('updates the move and player when a previous move is selected', () => {
  shared.wrapper.setState(winState);
  shared.wrapper.find('GameInfo').props().onMoveClick(0);
  expect(shared.wrapper.state().move).toEqual(0);
  expect(shared.wrapper.state().player).toEqual(shared.initialState.player);
});

it('should select the correct player when a move is selected', () => {
  shared.wrapper.setState(winState);
  shared.wrapper.find('GameInfo').props().onMoveClick(1);
  expect(shared.wrapper.state().move).toEqual(1);
  expect(shared.wrapper.state().player).not.toEqual(shared.initialState.player);
});

it('should correctly calculate a draw', () => {
  shared.wrapper.setState(drawState);
  expect(shared.wrapper.getElement()).toMatchSnapshot();
});

it('should correctly calculate a win', () => {
  shared.wrapper.setState(winState);
  expect(shared.wrapper.getElement()).toMatchSnapshot();
});

it('should output the winning sequence', () => {
  shared.wrapper.setState(winState);
  expect(shared.wrapper.find('Board').props().winner.toJS()).toEqual([0, 4, 8]);
});

it('should not update when a previously selected square is selected', () => {
  shared.wrapper.find('Board').props().onClick(4, 1, 3);
  const interimState = shared.wrapper.state();
  shared.wrapper.find('Board').props().onClick(4, 1, 3);
  expect(shared.wrapper.state()).toEqual(interimState);
});

it('should not update when the game is finished and a square is selected', () => {
  shared.wrapper.setState(winState);
  shared.wrapper.find('Board').props().onClick(0, 1, 4);
  expect(shared.wrapper.state()).toEqual(winState);
});
