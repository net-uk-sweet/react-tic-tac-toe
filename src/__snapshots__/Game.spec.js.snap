// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="game"
>
  <div
    className="game-board"
  >
    <Board
      columns={3}
      onClick={[Function]}
      squares={
        Immutable.List [
          Immutable.Map {
            token: null,
            column: null,
            row: null,
          },
          Immutable.Map {
            token: null,
            column: null,
            row: null,
          },
          Immutable.Map {
            token: null,
            column: null,
            row: null,
          },
          Immutable.Map {
            token: null,
            column: null,
            row: null,
          },
          Immutable.Map {
            token: null,
            column: null,
            row: null,
          },
          Immutable.Map {
            token: null,
            column: null,
            row: null,
          },
          Immutable.Map {
            token: null,
            column: null,
            row: null,
          },
          Immutable.Map {
            token: null,
            column: null,
            row: null,
          },
          Immutable.Map {
            token: null,
            column: null,
            row: null,
          },
        ]
      }
      winner={null}
    />
  </div>
  <div
    className="game-panel"
  >
    <GameInfo
      ascendingOrder={true}
      currentMove={0}
      history={
        Immutable.List [
          Immutable.Map {
            squares: Immutable.List [
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
            ],
          },
        ]
      }
      onMoveClick={[Function]}
      status="Next player: x"
    />
    <button
      className="button qa-toggle-button"
      onClick={[Function]}
    >
      Toggle order
    </button>
    <button
      className="button qa-reset-button"
      onClick={[Function]}
    >
      Reset
    </button>
  </div>
</div>
`;

exports[`should correctly calculate a draw 1`] = `
<div
  className="game"
>
  <div
    className="game-board"
  >
    <Board
      columns={3}
      onClick={[Function]}
      squares={
        Immutable.List [
          Immutable.Map {
            token: "x",
            row: 1,
            column: 1,
          },
          Immutable.Map {
            token: "o",
            row: 1,
            column: 2,
          },
          Immutable.Map {
            token: "x",
            row: 1,
            column: 3,
          },
          Immutable.Map {
            token: "x",
            row: 2,
            column: 1,
          },
          Immutable.Map {
            token: "o",
            row: 2,
            column: 2,
          },
          Immutable.Map {
            token: "x",
            row: 2,
            column: 3,
          },
          Immutable.Map {
            token: "o",
            row: 3,
            column: 1,
          },
          Immutable.Map {
            token: "x",
            row: 3,
            column: 2,
          },
          Immutable.Map {
            token: "o",
            row: 3,
            column: 3,
          },
        ]
      }
      winner={null}
    />
  </div>
  <div
    className="game-panel"
  >
    <GameInfo
      ascendingOrder={true}
      currentMove={9}
      history={
        Immutable.List [
          Immutable.Map {
            squares: Immutable.List [
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
            ],
          },
          Immutable.Map {
            index: 0,
            squares: Immutable.List [
              Immutable.Map {
                token: "x",
                row: 1,
                column: 1,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
            ],
          },
          Immutable.Map {
            index: 1,
            squares: Immutable.List [
              Immutable.Map {
                token: "x",
                row: 1,
                column: 1,
              },
              Immutable.Map {
                token: "o",
                row: 1,
                column: 2,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
            ],
          },
          Immutable.Map {
            index: 3,
            squares: Immutable.List [
              Immutable.Map {
                token: "x",
                row: 1,
                column: 1,
              },
              Immutable.Map {
                token: "o",
                row: 1,
                column: 2,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: "x",
                row: 2,
                column: 1,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
            ],
          },
          Immutable.Map {
            index: 4,
            squares: Immutable.List [
              Immutable.Map {
                token: "x",
                row: 1,
                column: 1,
              },
              Immutable.Map {
                token: "o",
                row: 1,
                column: 2,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: "x",
                row: 2,
                column: 1,
              },
              Immutable.Map {
                token: "o",
                row: 2,
                column: 2,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
            ],
          },
          Immutable.Map {
            index: 7,
            squares: Immutable.List [
              Immutable.Map {
                token: "x",
                row: 1,
                column: 1,
              },
              Immutable.Map {
                token: "o",
                row: 1,
                column: 2,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: "x",
                row: 2,
                column: 1,
              },
              Immutable.Map {
                token: "o",
                row: 2,
                column: 2,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: "x",
                row: 3,
                column: 2,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
            ],
          },
          Immutable.Map {
            index: 6,
            squares: Immutable.List [
              Immutable.Map {
                token: "x",
                row: 1,
                column: 1,
              },
              Immutable.Map {
                token: "o",
                row: 1,
                column: 2,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: "x",
                row: 2,
                column: 1,
              },
              Immutable.Map {
                token: "o",
                row: 2,
                column: 2,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: "o",
                row: 3,
                column: 1,
              },
              Immutable.Map {
                token: "x",
                row: 3,
                column: 2,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
            ],
          },
          Immutable.Map {
            index: 5,
            squares: Immutable.List [
              Immutable.Map {
                token: "x",
                row: 1,
                column: 1,
              },
              Immutable.Map {
                token: "o",
                row: 1,
                column: 2,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: "x",
                row: 2,
                column: 1,
              },
              Immutable.Map {
                token: "o",
                row: 2,
                column: 2,
              },
              Immutable.Map {
                token: "x",
                row: 2,
                column: 3,
              },
              Immutable.Map {
                token: "o",
                row: 3,
                column: 1,
              },
              Immutable.Map {
                token: "x",
                row: 3,
                column: 2,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
            ],
          },
          Immutable.Map {
            index: 8,
            squares: Immutable.List [
              Immutable.Map {
                token: "x",
                row: 1,
                column: 1,
              },
              Immutable.Map {
                token: "o",
                row: 1,
                column: 2,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: "x",
                row: 2,
                column: 1,
              },
              Immutable.Map {
                token: "o",
                row: 2,
                column: 2,
              },
              Immutable.Map {
                token: "x",
                row: 2,
                column: 3,
              },
              Immutable.Map {
                token: "o",
                row: 3,
                column: 1,
              },
              Immutable.Map {
                token: "x",
                row: 3,
                column: 2,
              },
              Immutable.Map {
                token: "o",
                row: 3,
                column: 3,
              },
            ],
          },
          Immutable.Map {
            index: 2,
            squares: Immutable.List [
              Immutable.Map {
                token: "x",
                row: 1,
                column: 1,
              },
              Immutable.Map {
                token: "o",
                row: 1,
                column: 2,
              },
              Immutable.Map {
                token: "x",
                row: 1,
                column: 3,
              },
              Immutable.Map {
                token: "x",
                row: 2,
                column: 1,
              },
              Immutable.Map {
                token: "o",
                row: 2,
                column: 2,
              },
              Immutable.Map {
                token: "x",
                row: 2,
                column: 3,
              },
              Immutable.Map {
                token: "o",
                row: 3,
                column: 1,
              },
              Immutable.Map {
                token: "x",
                row: 3,
                column: 2,
              },
              Immutable.Map {
                token: "o",
                row: 3,
                column: 3,
              },
            ],
          },
        ]
      }
      onMoveClick={[Function]}
      status="Draw"
    />
    <button
      className="button qa-toggle-button"
      onClick={[Function]}
    >
      Toggle order
    </button>
    <button
      className="button qa-reset-button"
      onClick={[Function]}
    >
      Reset
    </button>
  </div>
</div>
`;

exports[`should correctly calculate a win 1`] = `
<div
  className="game"
>
  <div
    className="game-board"
  >
    <Board
      columns={3}
      onClick={[Function]}
      squares={
        Immutable.List [
          Immutable.Map {
            token: "x",
            row: 1,
            column: 1,
          },
          Immutable.Map {
            token: "o",
            row: 1,
            column: 2,
          },
          Immutable.Map {
            token: null,
            column: null,
            row: null,
          },
          Immutable.Map {
            token: null,
            column: null,
            row: null,
          },
          Immutable.Map {
            token: "x",
            row: 2,
            column: 2,
          },
          Immutable.Map {
            token: null,
            column: null,
            row: null,
          },
          Immutable.Map {
            token: "o",
            row: 3,
            column: 1,
          },
          Immutable.Map {
            token: null,
            column: null,
            row: null,
          },
          Immutable.Map {
            token: "x",
            row: 3,
            column: 3,
          },
        ]
      }
      winner={
        Immutable.List [
          0,
          4,
          8,
        ]
      }
    />
  </div>
  <div
    className="game-panel"
  >
    <GameInfo
      ascendingOrder={true}
      currentMove={5}
      history={
        Immutable.List [
          Immutable.Map {
            squares: Immutable.List [
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
            ],
          },
          Immutable.Map {
            index: 0,
            squares: Immutable.List [
              Immutable.Map {
                token: "x",
                row: 1,
                column: 1,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
            ],
          },
          Immutable.Map {
            index: 1,
            squares: Immutable.List [
              Immutable.Map {
                token: "x",
                row: 1,
                column: 1,
              },
              Immutable.Map {
                token: "o",
                row: 1,
                column: 2,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
            ],
          },
          Immutable.Map {
            index: 4,
            squares: Immutable.List [
              Immutable.Map {
                token: "x",
                row: 1,
                column: 1,
              },
              Immutable.Map {
                token: "o",
                row: 1,
                column: 2,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: "x",
                row: 2,
                column: 2,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
            ],
          },
          Immutable.Map {
            index: 6,
            squares: Immutable.List [
              Immutable.Map {
                token: "x",
                row: 1,
                column: 1,
              },
              Immutable.Map {
                token: "o",
                row: 1,
                column: 2,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: "x",
                row: 2,
                column: 2,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: "o",
                row: 3,
                column: 1,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
            ],
          },
          Immutable.Map {
            index: 8,
            squares: Immutable.List [
              Immutable.Map {
                token: "x",
                row: 1,
                column: 1,
              },
              Immutable.Map {
                token: "o",
                row: 1,
                column: 2,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: "x",
                row: 2,
                column: 2,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: "o",
                row: 3,
                column: 1,
              },
              Immutable.Map {
                token: null,
                column: null,
                row: null,
              },
              Immutable.Map {
                token: "x",
                row: 3,
                column: 3,
              },
            ],
          },
        ]
      }
      onMoveClick={[Function]}
      status="Winner: x"
    />
    <button
      className="button qa-toggle-button"
      onClick={[Function]}
    >
      Toggle order
    </button>
    <button
      className="button qa-reset-button"
      onClick={[Function]}
    >
      Reset
    </button>
  </div>
</div>
`;
